Convert md to Ecmarkup

bterlson opened the pull request:
This PR converts the md spec documents to Ecmarkup. The built output can be [previewed here](http://bterlson.github.io/ecmascript-object-observe/). More information about Ecmarkup, including basic instructions for how to build locally, can be found at [http://bterlson.github.io/ecmarkup](http://bterlson.github.io/ecmarkup). Auto-updating the gh-pages branch can be done with Travis-ci - Domenic created [some instructions](https://gist.github.com/domenic/ec8b0fc8ab45f39403dd).

domenic commented on KeyAlgorithmsAndSemantics.html:
```diff
@@ -0,0 +1,71 @@
+    <h1>ChangeObserver and PendingChangeRecords</h1>
+    <p>Conceptually, at the heart of the EMCAScript observation mechanism is two structures:</p>
+    <ul>
+      <li><a href="#ChangeObservers">ChangeObservers</a>: the set of functions which are presently registered as observers of changes which take place to given object.
```
This file appears to be output instead of input?

bterlson replied:
No. This is a bulleted list, not algorithm steps, hence vanilla html lists. Maybe ecmarkdown's paragraph function should handle bulleted lists?

domenic replied:
Nevermind, I understand it better now :)

domenic commented on NewInternalsSpecification.html:
```diff
@@ -0,0 +1,271 @@
+    </emu-alg>
+  </emu-clause>
+
+  <emu-clause id="ShouldDeliverToObserver" aoid="ShouldDeliverToObserver">
```
Hmm this duplication seems kind of annoying for authors. Unsure of the solution though.

bterlson replied:
I agree. It's not a good idea to auto-link based on id since most clauses aren't abstract operations. However, I think it's reasonable to say if you have an aoid with an empty value that it will pull from the id.

bterlson replied:
Created an issue for this: https://github.com/bterlson/ecmarkup/issues/5.

domenic commented on NewInternalsSpecification.html:
```diff
@@ -0,0 +1,271 @@
+          1. If _changeObservers_ is empty, return.
+          1. Let _target_ be the value of _O_'s [[Target]] internal slot.
+          1. Let _newRecord_ be ObjectCreate(%ObjectPrototype%).
+          1. Let _desc_ be the PropertyDescriptor{[[Value]]: _target_, [[Writable]]: *false*, [[Enumerable]]: *true*, [[Configurable]]: *false*}.
```
Stylesheet needs to bold emu-vals. They are outputting correctly but not visually distinct.

bterlson replied:
Done.

domenic commented on NewInternalsSpecification.html:
```diff
@@ -0,0 +1,271 @@
+      1. Return *true*.
+    </emu-alg>
+
+    <emu-note>The user facing function ''Object.deliverChangeRecords'' returns *''undefined''* to prevent detection if anything was delivered or not.</emu-note>
```
These ''s seem unnecessary and strange?

bterlson replied:
Agreed, those were in the source document, not really sure what they're for. Looks like I can just replace with single back-ticks for the first and just * for the second.

arv replied:
The ecma wiki used a different wiki dialect. I haven't gone through all the documents to use markdown dialect.

One of the benefits of md was that I could edit the md on github.org. I'm not sure what to do about that though.

domenic commented on NewInternalsSpecification.html:
```diff
@@ -0,0 +1,271 @@
+        1. Let _success_ be the result of calling the [[DefineOwnProperty]] internal method of _changeRecord_ passing `"name"` and _desc_ as arguments.
+        1. Assert: _success_ is *true*.
+      1. If IsDataDescriptor(_oldDesc_) is *true*, then
+        1. If IsDataDescritor(_newDesc_) is *false* or SameValue(_oldDesc_.[[Value]], _newDesc_.[[Value]]) is *false*, then
```
Typo "Descritor" (in the original text too)

arv replied:
Brian, ignore these typos. I'll take care of these later.

domenic commented on NewInternalsSpecification.html:
```diff
@@ -0,0 +1,271 @@
+    <p>When the abstract operation CreateSpliceChangeRecord is called with the arguments: _object_, _index_, _removed_, and _addedCount_, the following steps are taken:</p>
+    <emu-alg>
+      1. Let _changeRecord_ be the result of the abstraction operation ObjectCreate (15.2).
+      1. Call the [[DefineOwnProperty]] internal method of _changeRecord_ with arguments *''"type"''*, Property Descriptor {[[Value]]: "splice", [[Writable]]: *false*, [[Enumerable]]: *true*, [[Configurable]]: *false*}, and *false*.
```
More weird single quotes

domenic commented on ChangesToTheExistingSpecification.html:
```diff
@@ -0,0 +1,498 @@
+        1. Assert: _extensible_ is *true*.
+        1. If IsGenericDescriptor(_Desc_) or IsDataDescriptor(_Desc_) is *true*, then
+          1. If _O_ is not *undefined*, then create an own data property named _P_ of object _O_ whose [[Value]], [[Writable]], [[Enumerable]] and [[Configurable]] attribute values are described by Desc. If the value of an attribute field of Desc is absent, the attribute of the newly created property is set to its default value.
+        1. Else Desc must be an accessor Property Descriptor,
```
This algorithm in general is missing some punctuation; Desc and current need `_`s, and true/false need `*`s

domenic commented on ChangesToTheExistingSpecification.html:
```diff
@@ -0,0 +1,498 @@
+      1. If _desc_.[[Configurable]] is *true*, then
+        1. Remove the own property with name _P_ from _O_.
+        1. <ins>Let _R_ be CreateChangeRecord(`"delete"`, _O_, _P_, _desc_).</ins>
+        1. <ins>Call EnqueueChangeRecord(_O_ and _R_).</ins>
```
"and"!?

domenic commented:
I realized as I was reviewing probably a lot of the minor critiques are just from the original markdown. And indeed they are. So I'll stop doing that :). The extra `'`s seems worth fixing though.

arv commented on KeyAlgorithmsAndSemantics.html:
```diff
@@ -0,0 +1,71 @@
+
+    <p>An observer must only receive change records whose type matches one provided in the accept list, which is passed as the (optional) third argument to <a href="#object.observe">Object.observe</a> (if the argument is not provided, the list defaults to the "intrinsic" set of change types).</p>
+
+    <p>An observation is conceptually a tuple: <observed object, observer function, accepted change types>. Object.observe ensures that this tuple is represented as an observerRecord in the associated Notifier's ChangeObservers list.</p>
```
We need to use `&lt;` here now.

arv commented:
LGTM. I'll merge this on Monday and set up a travis-ci script to auto push.

bterlson commented:
Implemented the aoid default and syntax highlighting. Updated the PR. Preview link is [the same](http://bterlson.github.io/ecmascript-object-observe/).

bterlson commented:
:+1: 

Here's Ecmarkup's [.travis.yaml file](https://github.com/bterlson/ecmarkup/blob/master/.travis.yml) if it helps with your travis setup.
